SparseGrids/tsgGridWavelet.cpp

```diff
@@ -331,6 +366,32 @@ double GridWavelet::evalIntegral(const int p[]) const{
     }
     return v;
 }
+void GridWavelet::evalDiffBasis(const int p[], const double x[], double jacobian[]) const {
+    // Evaluates the derivative of the wavelet basis given at point p at the coordinates given by x.
+    // The derivative function of a symmetric function will be antisymmetric across the origin for every x != 0.0.
+    std::vector<double> value_cache(num_dimensions), derivative_cache(num_dimensions);
+    for(int i=0; i<num_dimensions; i++) {
+        value_cache[i] = rule1D.eval<0>(p[i], x[i]);
+        derivative_cache[i] = rule1D.eval<1>(p[i], x[i]);
+    }
+
+    std::fill(jacobian, jacobian + num_dimensions, 0.0);
+    // Note that excessive branching is done to mirror the logic in evalBasis() and evalIntegral().
+    for(int i=0; i<num_dimensions; i++) {
```

Comment:
You have nested loops that perform `num_dimensions^2` operations and even with the if-statements and the breaks, the complexity is bounded below by `num_dimensions`.

You can guarantee complexity of `num_dimensions` with the following:
```
double t = 1.0;
for(int i=1; i<num_dimensions; i++){
    t *= value_cache[i];
    jacobian[i] *= t;
}
t = 1.0;
for(int i=num_dimensions-1; i>=0; i--){
    t *= value_cache[i];
    jacobian[i] *= t;
}
```
We can break the first loop when `t` becomes zero and zero out the remainder of the jacobian and we can shorten the second loop to only the non-negative portions (and zero out if `t` hits zero), but the zero-out will still require operations and the current set assumes that we don't have `derivative_cache`, instead the derivative values are already written into the Jacobian.

Reply:
Sounds good. I changed this in the latest push to include the zero-ing out logic. 

I do think that it should be `value_cache[i-1]` in the first block, `value_cache[i+1]` in the second block, and `for(int i=num_dimensions-2; i>=0; i--)` for the second for-loop, though.